pkg/apiserver/utils/stats/clickhouse_stats.go

```diff
@@ -0,0 +1,195 @@
+	c := &ClickHouseStatQuerierImpl{
+		kubeClient: kubeClient,
+	}
+	return c
+}
+
+func (c *ClickHouseStatQuerierImpl) GetDiskInfo(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(diskQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting diskInfo from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) GetTableInfo(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(tableInfoQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting tableInfo from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) GetInsertRate(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(insertRateQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting insertRate from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) GetStackTrace(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(stackTraceQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting stackTrace from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) getDataFromClickHouse(query int, namespace string, stats *v1alpha1.ClickHouseStats) error {
+	var err error
+	if c.clickhouseConnect == nil {
+		c.clickhouseConnect, err = util.SetupClickHouseConnection(c.kubeClient, namespace)
+		if err != nil {
+			return err
+		}
+	}
+	result, err := c.clickhouseConnect.Query(queryMap[query])
+	if err != nil {
+		c.clickhouseConnect = nil
+		return fmt.Errorf("failed to get data from clickhouse: %v", err)
+	}
+	defer result.Close()
+	for result.Next() {
+		var err error
+		switch query {
+		case diskQuery:
+			res := v1alpha1.DiskInfo{}
+			err = result.Scan(&res.Shard, &res.Database, &res.Path, &res.FreeSpace, &res.TotalSpace, &res.UsedPercentage)
+			res.UsedPercentage = res.UsedPercentage + " %"
+			stats.DiskInfos = append(stats.DiskInfos, res)
+		case tableInfoQuery:
+			res := v1alpha1.TableInfo{}
+			var totalRows sql.NullString
+			var totalBytes sql.NullString
+			err = result.Scan(&res.Shard, &res.Database, &res.TableName, &totalRows, &totalBytes, &res.TotalCols)
+			if !totalRows.Valid || !totalBytes.Valid {
+				continue
+			}
+			res.TotalRows = totalRows.String
+			res.TotalBytes = totalBytes.String
+			stats.TableInfos = append(stats.TableInfos, res)
+		case insertRateQuery:
+			res := v1alpha1.InsertRate{}
+			err = result.Scan(&res.Shard, &res.RowsPerSec, &res.BytesPerSec)
+			stats.InsertRates = append(stats.InsertRates, res)
+		case stackTraceQuery:
+			res := v1alpha1.StackTrace{}
+			err = result.Scan(&res.Shard, &res.TraceFunctions, &res.Count)
+			stats.StackTraces = append(stats.StackTraces, res)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to parse the data returned by database: %v", err)
```

Comment:
This is just to make a little less ignorant... can this query return multiple rows?

In that case, would it better to skip and move to the next record if we fail to parse a record?

Reply:
The query returns multiple rows.
I think it is a good idea to skip the error row instead of throwing away the whole data?
cc @yanjunz97 and @wsquan171 for some feedbacks.

Reply:
Agree to skip the error row. It would be better if we could display the results retrieved successfully and the error regarding to the error row at the same time?

Reply:
I added a ErrorMsg, which is a []string, in ClickHouseStats. If we have an error row, we will skip it and add the error msg to ErrorMsg. On CLI side, it will show both ErrorMsgs and the results.

Reply:
The new approach looks good to me.